### PR TITLE
feat(api): Do not return receipt if tx was not included to the batch

### PR DESCRIPTION
**Pull request by Deniallugo**
## What ❔

For compatibility with Ethereum, we have to return `Receipt` only if tx is included in the miniblock. Block hash is not null null. 

## Why ❔

For compatibility with Ethereum and for not breaking external clients, such as Metemask, they do not expect that block hash could be null

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `cargo spellcheck --cfg=./spellcheck/era.cfg --code 1`.

### CODE REVIEW

File: core/lib/types/src/api/mod.rs [hunk 1]
```diff
@@ -209,7 +209,7 @@ pub struct TransactionReceipt {
     pub transaction_index: Index,
     /// Hash of the block this transaction was included within.
     #[serde(rename = "blockHash")]
-    pub block_hash: Option<H256>,
+    pub block_hash: H256,
```
Comment (perekopskiy) [inline core/lib/types/src/api/mod.rs:212]: If you're changing the struct then it also makes sense to make `block_number`, `status` required

**Review by perekopskiy — DISMISSED** [review]

File: core/lib/dal/src/transactions_web3_dal.rs [hunk 2]
```diff
@@ -78,21 +78,17 @@ impl TransactionsWeb3Dal<'_, '_> {
             .fetch_optional(self.storage.conn())
             .await?
             .map(|db_row| {
-                let status = match (db_row.block_number, db_row.error) {
-                    (_, Some(_)) => Some(U64::from(0)),
-                    (Some(_), None) => Some(U64::from(1)),
-                    // tx not executed yet
-                    _ => None,
-                };
+                let status = db_row.error.map(|_| U64::zero()).unwrap_or_else(U64::one);
```
Comment (perekopskiy) [inline core/lib/dal/src/transactions_web3_dal.rs:81]: unwrap_or_else(U64::one) looks weirds, it should be smth like `expect("Mined transaction must have status")`

File: core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs [hunk 3]
```diff
@@ -523,9 +523,7 @@ impl<G: L1GasPriceProvider> EthNamespace<G> {
                     .await
                     .map_err(|err| internal_error(METHOD_NAME, err))
                 {
-                    if main_node_receipt.status == Some(0.into())
-                        && main_node_receipt.block_number.is_none()
-                    {
+                    if main_node_receipt.status == 0.into() {
```
Comment (perekopskiy) [inline core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs:526]: Given that we don't return receipts for rejected transactions anymore, forwarding requests to main node doesn't make sense anymore.

File: core/lib/dal/src/transactions_web3_dal.rs (hunk 2 above)
Comment (Deniallugo) [reply core/lib/dal/src/transactions_web3_dal.rs:81]: no, we unwrap the error.

Comment (Deniallugo) [reply core/lib/dal/src/transactions_web3_dal.rs:81]: it's just `if none then zero else one  `

**Review by RomanBrodetski — DISMISSED** [review]

**Review by perekopskiy — APPROVED** [review]

**Review by RomanBrodetski — APPROVED** [review]

**Review by RomanBrodetski — APPROVED** [review]

**Pull request merged**